compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp

```diff
@@ -0,0 +1,423 @@
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include <sstream>
+
+#define DEBUG_TYPE "iree-global-opt-lift-generic-to-tranpose-batch-matmul"
+
+namespace mlir {
+namespace iree_compiler {
+namespace GlobalOptimization {
+
+namespace {
+
+template <typename T, unsigned int N>
+mlir::raw_ostream &operator<<(mlir::raw_ostream &s,
+                              const SmallVector<T, N> &v) {
+  s << "[ ";
+  for (const auto &e : v) {
+    s << e << " ";
+  }
+  s << "]";
+  return s;
+}
+
+bool isCastOfBlockArgument(Operation *op) {
+  return isa<CastOpInterface>(op) && op->getNumOperands() == 1 &&
+         isa<BlockArgument>(op->getOperand(0));
+}
+
+bool isCastOrInputBlockArgument(Value input, int64_t numInputs) {
+  if (!input.isa<BlockArgument>()) {
+    Operation *castOp0 = input.getDefiningOp();
+    if (!castOp0 || !isCastOfBlockArgument(castOp0)) {
+      return false;
+    }
+    return castOp0->getOperand(0).cast<BlockArgument>().getArgNumber() !=
+           numInputs;
+  } else {
+    return input.cast<BlockArgument>().getArgNumber() != numInputs;
+  }
+}
+
+// Returns true if the linalg::GenericOp has a body like a matmul. This
+// does not check the indexing maps
+//
+// This function looks for a body like:
+// ```mlir
+// ^bb0(%in: !lhs, %in_0: !rhs, %out: !out):
+//   %3 = arith.extui %in : !lhs to !out
+//   %4 = arith.extsi %in_0 : !rhs to !out
+//   %5 = arith.muli %3, %4 : !out
+//   %6 = arith.addi %5, %out : !out
+//   linalg.yield %6 : !out
+// ```
+// Ensuring the following conditions:
+//    1) linalg.yield result comes from an arith.add op, accumulating on %out
+//    2) The other arith.add operand comes from arith.mul
+//    3) Both arith.mul operands are either block input arguments, or produced
+//       by a `CastOpInterface` of a block input argument
+static bool hasMatmulBody(linalg::GenericOp genericOp) {
+  int numInputs = genericOp.getNumDpsInputs();
+  if (numInputs != 2) {
+    LLVM_DEBUG(llvm::dbgs() << "op does not have exactly 2 inputs\n");
+    return false;
+  }
+  int numOutputs = genericOp.getNumDpsInits();
+  if (numOutputs != 1) {
+    LLVM_DEBUG(llvm::dbgs() << "op does not have exactly 1 output\n");
+    return false;
+  }
+
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+  Value yieldedValue = yieldOp->getOperand(0);
+
+  // Check that yielded value is an arith.add op, and is accumulating
+  Operation *addOp = yieldedValue.getDefiningOp();
+  if (!addOp) {
+    LLVM_DEBUG(llvm::dbgs() << "linalg.yield operand has no defining op\n");
+    return false;
+  }
+  if (!isa<arith::AddFOp>(*addOp) && !isa<arith::AddIOp>(*addOp)) {
```

Comment:
nit: you can merge them into a single `isa`.

```suggestion
  if (!isa<arith::AddFOp, arith::AddIOp>(*addOp)) {
```